### ISSUE TITLE
`spago update` command should work with solver-packages too

### ISSUE THREAD

**Issue by f-f** [purescript/spago#1001]
When the project build happens with the solver (rather than package sets), we should:
1. try to build a plan with the latest versions on the Registry that are still compatible with the current compiler
2. build with that plan
3. and if all is well put the new ranges in the config file

**Comment by f-f** [comment]
Pivoting this issue to solver-projects only, since the package sets side is done

**Comment by f-f** [comment]
I think the solver will try to pick latest dependencies first, so if we just throw away the current ranges and put in the widest ranges then we can just ask the solver to give us a plan.

I think this is blocked by #1003, because we should update the lockfile when the plan is good - right now we don't use the lockfile in the right way, and there are no tests for it.

**Comment by flip111** [comment]
label size:medium seems more applicable to me. Or perhaps somebody can provide some pointers to this issue?

**Comment by f-f** [comment]
What I really mean with the `size:small` label is "under 50 lines" (this one might be slightly bigger than that, but the spec is very clear so I think it's affordable) - some patches are 5-10 lines, and that would deserve a `size:micro` label 😄

A few pointers for this:
- the unimplemented branch is here, so you can replace this error with the new code:
  https://github.com/purescript/spago/blob/b450892d69aa3013861fde6c73299d2c17aba3dc/src/Spago/Command/Upgrade.purs#L26
- upgrading the package set is about the whole workspace, but for upgrading solver ranges we want to deal with a single package/config file, so we'll want to select a single package like this:
  https://github.com/purescript/spago/blob/b450892d69aa3013861fde6c73299d2c17aba3dc/bin/src/Main.purs#L753-L767
- at this point you can look inside this `selected :: WorkspacePackage`, and get its `Dependencies`. Hold on to them because you'll need them later (e.g. in a `currentDependencies` variable)
- then create new ranges for the solver run, where you take the `currentDependencies`, and just assign the `Config.widestRange` to every package
- ...so you can call the solver to get a new plan:
  https://github.com/purescript/spago/blob/b450892d69aa3013861fde6c73299d2c17aba3dc/src/Spago/Command/Fetch.purs#L460
- if the plan is bad then fail with the error, but if it's good we can overwrite the current ranges in the config file, as we do for the `ensureRanges` flag (note the call to `getRangeFromPackage`, we'll want to union the old `Range` in `currentRanges` with the new `Range` that we get from this function):
  https://github.com/purescript/spago/blob/b450892d69aa3013861fde6c73299d2c17aba3dc/src/Spago/Command/Fetch.purs#L162-L170

**Comment by fsoikin** [comment]
I don't understand the "union" step. Wouldn't it negate all the work the solver has done? Like, if I add extra versions to the range, they might not be compatible with other packages, no? Or did you mean "intersect" instead of "union"?

**Comment by f-f** [comment]
@fsoikin what I meant above is that we should not inadvertently introduce a breaking change in version ranges - e.g. if your project `a` is a library that is currently depending on `foo` at versions `>=3.0.0 <4.0.0` and the update moves its range to `>=4.0.0 <5.0.0`, then a project `b` that depends on `a` and `foo@3.1.4` won't be compatible with the new version of `a`.

So I mentioned unioning the ranges because then we'd get a range `>=3.0.0 <5.0.0` which would not be breaking. However:
- we would still have to validate the new ranges with the solver
- I am not sure if unioning them is the right thing to do and it's not going to be correct all the time, but the starting point is that the solver will give us a set of exact versions and not a range. We'd then need to make it a range, which is a, uhh, "creative" operation, so we need some heuristics here
- ultimately we don't have to do a perfect job - the baseline we need to beat is someone updating dependencies by hand in the config file. As long as we can make that job less tedious while having approximately the same amount of correctness then we're good I think
